# ДЗ к Урок 2. Создание признакового пространства 

Продолжим обработку данных с Твиттера.

In [1]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd

combine_df=pd.read_pickle('combine_df.pkl')
combine_df.head()


id  label                                              tweet  \
0   1    0.0   @user when a father is dysfunctional and is s...   
1   2    0.0  @user @user thanks for #lyft credit i can't us...   
2   3    0.0                                bihday your majesty   
3   4    0.0  #model   i love u take with u all the time in ...   
4   5    0.0             factsguide: society now    #motivation   

                                         clean_tweet  \
0  when father is dysfunctional and is so selfish...   
1  thanks for lyft credit cannot use cause they d...   
2                                bihday your majesty   
3  model love you take with you all the time in your   
4                  factsguide society now motivation   

                                         tweet_token  \
0  [when, father, is, dysfunctional, and, is, so,...   
1  [thanks, for, lyft, credit, can, not, use, cau...   
2                            [bihday, your, majesty]   
3  [model, love, you, take, with, you, all, the, ...   
4             [factsguide, society, now, motivation]   

                                tweet_token_filtered  \
0  [father, dysfunctional, selfish, drags, kids, ...   
1  [thanks, lyft, credit, use, cause, offer, whee...   
2                                  [bihday, majesty]   
3                          [model, love, take, time]   
4                  [factsguide, society, motivation]   

                                       tweet_stemmed  \
0  [father, dysfunct, selfish, drag, kid, dysfunc...   
1  [thank, lyft, credit, use, caus, offer, wheelc...   
2                                  [bihday, majesti]   
3                          [model, love, take, time]   
4                        [factsguid, societi, motiv]   

                                    tweet_lemmatized  
0  [father, dysfunctional, selfish, drag, kid, dy...  
1  [thank, lyft, credit, use, cause, offer, wheel...  
2                                  [bihday, majesty]  
3                          [model, love, take, time]  
4                  [factsguide, society, motivation]

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.

    • Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
    
    • Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
    
    • Исключим стоп-слова с помощью stop_words='english'.
    
    • Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

In [2]:
from sklearn.feature_extraction.text import CountVectorizer

count_vectorizer = CountVectorizer(max_df=0.9, max_features = 1000, stop_words='english')

In [3]:
# Создаем the Bag-of-Words модель
bag_of_words_stemmed = count_vectorizer.fit_transform(combine_df['tweet_stemmed'].apply(lambda x: ' '.join(x)))

In [4]:
# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_stemmed.toarray(), columns = feature_names)

abl  absolut  accept  account  act  action  activ  actor  actual  ad  \
0        0        0       0        0    0       0      0      0       0   0   
1        0        0       0        0    0       0      0      0       0   0   
2        0        0       0        0    0       0      0      0       0   0   
3        0        0       0        0    0       0      0      0       0   0   
4        0        0       0        0    0       0      0      0       0   0   
...    ...      ...     ...      ...  ...     ...    ...    ...     ...  ..   
49154    0        0       0        0    0       0      0      0       0   0   
49155    0        0       0        0    0       0      0      0       0   0   
49156    0        0       0        0    0       0      0      0       0   0   
49157    0        0       0        0    0       0      0      0       0   0   
49158    0        0       0        0    0       0      0      0       0   0   

       ...  year  yesterday  yo  yoga  york  young  youtub  yoyou  yr  yummi  
0      ...     0          0   0     0     0      0       0      0   0      0  
1      ...     0          0   0     0     0      0       0      0   0      0  
2      ...     0          0   0     0     0      0       0      0   0      0  
3      ...     0          0   0     0     0      0       0      0   0      0  
4      ...     0          0   0     0     0      0       0      0   0      0  
...    ...   ...        ...  ..   ...   ...    ...     ...    ...  ..    ...  
49154  ...     0          0   0     0     0      0       0      0   0      0  
49155  ...     0          0   0     0     0      0       0      0   0      0  
49156  ...     0          0   0     0     0      0       0      0   0      0  
49157  ...     0          0   0     0     0      0       0      0   0      0  
49158  ...     0          0   0     0     0      0       0      0   0      0  

[49159 rows x 1000 columns]

In [5]:
# Создаем the Bag-of-Words модель
bag_of_words_lemmatized = count_vectorizer.fit_transform(combine_df['tweet_lemmatized'].apply(lambda x: ' '.join(x)))

In [6]:
# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_lemmatized.toarray(), columns = feature_names)

able  absolutely  accept  account  act  action  actor  actually  adapt  \
0         0           0       0        0    0       0      0         0      0   
1         0           0       0        0    0       0      0         0      0   
2         0           0       0        0    0       0      0         0      0   
3         0           0       0        0    0       0      0         0      0   
4         0           0       0        0    0       0      0         0      0   
...     ...         ...     ...      ...  ...     ...    ...       ...    ...   
49154     0           0       0        0    0       0      0         0      0   
49155     0           0       0        0    0       0      0         0      0   
49156     0           0       0        0    0       0      0         0      0   
49157     0           0       0        0    0       0      0         0      0   
49158     0           0       0        0    0       0      0         0      0   

       add  ...  yesterday  yo  yoga  york  young  youtube  yoyou  yr  yrs  \
0        0  ...          0   0     0     0      0        0      0   0    0   
1        0  ...          0   0     0     0      0        0      0   0    0   
2        0  ...          0   0     0     0      0        0      0   0    0   
3        0  ...          0   0     0     0      0        0      0   0    0   
4        0  ...          0   0     0     0      0        0      0   0    0   
...    ...  ...        ...  ..   ...   ...    ...      ...    ...  ..  ...   
49154    0  ...          0   0     0     0      0        0      0   0    0   
49155    0  ...          0   0     0     0      0        0      0   0    0   
49156    0  ...          0   0     0     0      0        0      0   0    0   
49157    0  ...          0   0     0     0      0        0      0   0    0   
49158    0  ...          0   0     0     0      0        0      0   0    0   

       yummy  
0          0  
1          0  
2          0  
3          0  
4          0  
...      ...  
49154      0  
49155      0  
49156      0  
49157      0  
49158      0  

[49159 rows x 1000 columns]

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.

    • Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
    
    • Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
    
    • Исключим стоп-слова с помощью stop_words='english'.
    
    • Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().

In [7]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(max_df=0.9, max_features = 1000, stop_words='english')

In [8]:
# Создаем the Bag-of-Words модель
bag_of_words_stemmed = tfidf_vectorizer.fit_transform(combine_df['tweet_stemmed'].apply(lambda x: ' '.join(x)))

In [9]:
# Отобразим Bag-of-Words модель как DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_stemmed.toarray(), columns = feature_names)

abl  absolut  accept  account  act  action  activ  actor  actual   ad  \
0      0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
1      0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
2      0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
3      0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
4      0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
...    ...      ...     ...      ...  ...     ...    ...    ...     ...  ...   
49154  0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
49155  0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
49156  0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
49157  0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   
49158  0.0      0.0     0.0      0.0  0.0     0.0    0.0    0.0     0.0  0.0   

       ...  year  yesterday   yo  yoga  york  young  youtub  yoyou   yr  yummi  
0      ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
1      ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
2      ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
3      ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
4      ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
...    ...   ...        ...  ...   ...   ...    ...     ...    ...  ...    ...  
49154  ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
49155  ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
49156  ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
49157  ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  
49158  ...   0.0        0.0  0.0   0.0   0.0    0.0     0.0    0.0  0.0    0.0  

[49159 rows x 1000 columns]

In [10]:
# Создаем the Bag-of-Words модель
bag_of_words_lemmatized = tfidf_vectorizer.fit_transform(combine_df['tweet_lemmatized'].apply(lambda x: ' '.join(x)))

In [11]:
# Отобразим Bag-of-Words модель как DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_lemmatized.toarray(), columns = feature_names)

able  absolutely  accept  account  act  action  actor  actually  adapt  \
0       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
1       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
2       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
3       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
4       0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
...     ...         ...     ...      ...  ...     ...    ...       ...    ...   
49154   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49155   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49156   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49157   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   
49158   0.0         0.0     0.0      0.0  0.0     0.0    0.0       0.0    0.0   

       add  ...  yesterday   yo  yoga  york  young  youtube  yoyou   yr  yrs  \
0      0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
1      0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
2      0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
3      0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
4      0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
...    ...  ...        ...  ...   ...   ...    ...      ...    ...  ...  ...   
49154  0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
49155  0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
49156  0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
49157  0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   
49158  0.0  ...        0.0  0.0   0.0   0.0    0.0      0.0    0.0  0.0  0.0   

       yummy  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
...      ...  
49154    0.0  
49155    0.0  
49156    0.0  
49157    0.0  
49158    0.0  

[49159 rows x 1000 columns]

3. Натренируем gensim.models.Word2Vec модель на наших данных.

    • Тренировать будем на токенизированных твитах combine_df['tweet_token']
    
    • Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
    
    • Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.

In [62]:
# !pip install gensim
from gensim.models import Word2Vec

In [78]:
# настроим и натренируем вектора слов на наших данных
w2v_model = Word2Vec(size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34)
w2v_model.build_vocab(combine_df['tweet_token'], progress_per=10000)
w2v_model.train(combine_df['tweet_token'], total_examples=w2v_model.corpus_count, epochs=20, report_delay=1)

(9185476, 11793780)

4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [79]:
# выведим слово наиболее близкое к 'dinner'
result = w2v_model.most_similar(positive=['dinner'])
print("{}: {:.4f}".format(*result[0]))

cookout: 0.6168


In [80]:
# выведим слово наиболее близкое к 'trump'
result = w2v_model.most_similar(positive=['trump'])
print("{}: {:.4f}".format(*result[0]))

donald: 0.5599


5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).

Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

In [83]:
w2v_model["food"]

array([ 0.011261  , -0.5883919 ,  0.29684368,  0.28026885,  0.30669293,
       -0.0806098 ,  0.12438712,  0.73898953, -0.60643286,  0.11475627,
        0.2051253 , -0.32696623,  0.22314627, -0.06314936,  0.05006584,
       -0.03573371, -0.04520864, -0.51906776, -0.33363566, -0.02172917,
        0.06281642,  0.22109261, -0.23729274,  0.03493248, -0.6170491 ,
        0.65495676,  0.07503269, -0.45572096,  0.16313292, -0.1628854 ,
       -0.67402405,  0.15832335, -0.17522603,  0.19772276, -0.6002375 ,
       -0.17407636,  0.10486861,  0.00141175,  0.40164766, -0.18564685,
        0.9185499 ,  0.42487663,  0.16666286,  0.03355553,  0.23897316,
        0.08832248, -0.3910469 , -0.4623211 , -0.56866467,  0.46430287,
        0.47541362, -0.36652806,  0.2221444 , -0.5636217 , -0.13012017,
       -0.23512317, -0.52242935, -0.02436601, -0.6063118 ,  0.23518951,
       -0.3246681 , -0.01502926, -0.33156833, -0.71371365, -0.11985461,
       -0.7677017 , -0.06478282,  0.937762  ,  1.1073506 , -0.01

6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.

Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200).

# Simple chat-bot example

In [93]:
# !pip install annoy
# !pip install stop_words
# !pip install pymorphy2
# !pip install pymorphy2-dicts
# !pip install DAWG-Python
import string
from pymorphy2 import MorphAnalyzer
from stop_words import get_stop_words
import annoy
from gensim.models import Word2Vec, FastText
import pickle
import numpy as np
from tqdm import tqdm_notebook

In [94]:
assert False

#Small preprocess of the answers

question = None
written = False

with open("prepared_answers.txt", "w") as fout:
    with open("Otvety.txt", "r") as fin:
        for line in tqdm_notebook(fin):
            if line.startswith("---"):
                written = False
                continue
            if not written and question is not None:
                fout.write(question.replace("\t", " ").strip() + "\t" + line.replace("\t", " "))
                written = True
                question = None
                continue
            if not written:
                question = line.strip()
                continue

AssertionError: 

In [95]:
def preprocess_txt(line):
    spls = "".join(i for i in line.strip() if i not in exclude).split()
    spls = [morpher.parse(i.lower())[0].normal_form for i in spls]
    spls = [i for i in spls if i not in sw and i != ""]
    return spls

In [ ]:
assert True

# Preprocess for models fitting

sentences = []

morpher = MorphAnalyzer()
sw = set(get_stop_words("ru"))
exclude = set(string.punctuation)
c = 0

with open("Otvety.txt", "r") as fin:
    for line in tqdm_notebook(fin):
        spls = preprocess_txt(line)
        sentences.append(spls)
        c += 1
        if c > 100000:
            break

In [16]:
sentences = [i for i in sentences if len(i) > 2]

In [25]:
sentences[0]

['вопрос', 'тдв', 'отдыхать', 'лично', 'советовать', 'завести']

In [17]:
modelW2V = Word2Vec(sentences=sentences, size=50, window=5, min_count=1)

In [18]:
modelFT = FastText(sentences=sentences, size=50, min_count=1, window=5)

In [19]:
w2v_index = annoy.AnnoyIndex(50 ,'angular')
ft_index = annoy.AnnoyIndex(50 ,'angular')

index_map = {}
counter = 0

with open("prepared_answers.txt", "r") as f:
    for line in tqdm_notebook(f):
        n_w2v = 0
        n_ft = 0
        spls = line.split("\t")
        index_map[counter] = spls[1]
        question = preprocess_txt(spls[0])
        
        vector_w2v = np.zeros(50)
        vector_ft = np.zeros(50)
        for word in question:
            if word in modelW2V:
                vector_w2v += modelW2V[word]
                n_w2v += 1
            if word in modelFT:
                vector_ft += modelFT[word]
                n_ft += 1
        if n_w2v > 0:
            vector_w2v = vector_w2v / n_w2v
        if n_ft > 0:
            vector_ft = vector_ft / n_ft
        w2v_index.add_item(counter, vector_w2v)
        ft_index.add_item(counter, vector_ft)
            
        counter += 1
        
        if counter > 100000:
            break

w2v_index.build(10)
ft_index.build(10)

True

In [20]:
w2v_index.get_nns_by_vector(np.zeros(50), 3)

[1873, 4179, 4232]

In [28]:
def get_response(question, index, model, index_map):
    question = preprocess_txt(question)
    vector = np.zeros(50)
    norm = 0
    for word in question:
        if word in model:
            vector += model[word]
            norm += 1
    if norm > 0:
        vector = vector / norm
    answers = index.get_nns_by_vector(vector, 3)
    return [index_map[i] for i in answers]

In [34]:
TEXT = "когда сменится президент"

In [35]:

get_response(TEXT, w2v_index, modelW2V, index_map)

['Скорее всего я или люди из моего окружения.Так что знакомтесь пока есть время и вожможность. \n',
 'Вам, мадам, к психиатру нужно. И как можно скорее.... \n',
 'Потому что большинство считали что это фарс чтобы подорвать репутацию новой власти Украины. Поэтому ему просто не сообщили место и день регистрации. И конечно потому что он был бы серьезным конкурентом на пост президента, его бы поддержала практически вся молодёжь. Но теперь он собирается стать мэром города Одессы, и в него есть все шансы. Дарт Вейдер выступает как интернет партия Украины, и ничего никому не обещает.. \n']

In [36]:
get_response(TEXT, ft_index, modelFT, index_map)

['Следите за новостями в Украине. Как только там начнётся - считайте, что караван верблюдов уже вышел.. . Ждите, когда он прийдёт к вам.. . А лучше - скупайте баксы.... \n',
 'Поторопился..... \n',
 'такого быть не может! если такое есть, это серьезное нарушение. тут можно и бабок срубить с ПНД. епт. \n']